File: backend/app/models/emg.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime
 
 
class EMGReading(BaseModel):
    timestamp: datetime
    emg_average: Optional[float] = None
    emg_rms: Optional[float] = None
    muscle_load: Optional[float] = None
 
 
class EMGStats(BaseModel):
    latest_emg_average: Optional[float]
    latest_emg_rms: Optional[float]
    latest_muscle_load: Optional[float]
    mean_emg_average: Optional[float]
    mean_emg_rms: Optional[float]
    mean_muscle_load: Optional[float]
    record_count: int
 
 
class EMGResponse(BaseModel):
    readings: List[EMGReading]
    stats: EMGStats
    channel_id: str
    fetched_at: datetime
 
 
# ── ML feature stubs ─────────────────────────────────────────────────────────
 
class FatigueAnalysis(BaseModel):
    fatigue_index: Optional[float] = None
    onset_detected: bool = False
    confidence: Optional[float] = None
 
 
class GestureLabel(BaseModel):
    label: Optional[str] = None
    confidence: Optional[float] = None
 
 
class TremorAnalysis(BaseModel):
    frequency_hz: Optional[float] = None
    amplitude: Optional[float] = None
    tremor_detected: bool = False

File: frontend/src/types/emg.ts
export interface EMGReading {
  timestamp: string; // ISO 8601 UTC
  emg_average: number | null;
  emg_rms: number | null;
  muscle_load: number | null;
}

export interface EMGStats {
  latest_emg_average: number | null;
  latest_emg_rms: number | null;
  latest_muscle_load: number | null;
  mean_emg_average: number | null;
  mean_emg_rms: number | null;
  mean_muscle_load: number | null;
  record_count: number;
}

export interface EMGResponse {
  readings: EMGReading[];
  stats: EMGStats;
  channel_id: string;
  fetched_at: string;
}

/** Flattened shape used by Recharts */
export interface ChartPoint {
  time: string;       // Local time label
  timestamp: number;  // Unix ms for sorting
  emg_average: number | null;
  emg_rms: number | null;
  muscle_load: number | null;
}
